File: Server/SmartSolarMicrogrid.Api/DTOs/Reservations/UpdateReservationRequest.cs
/*
 * SE4040 - Enterprise Application Development
 * Smart Solar Microgrid Trading System
 * File: UpdateReservationRequest.cs
 * Purpose: Receive client parameters for modifying a scheduled energy slot reservation.
 */

namespace SmartSolarMicrogrid.Api.DTOs.Reservations;

public sealed class UpdateReservationRequest
{
    public DateTime ReservationDateTime { get; set; }

    public string? SlotId { get; set; }

    public string? ReservationType { get; set; }
}

File: Android/app/src/main/java/com/example/smartsolarmicrogridtradingsystem/feature/reservationqr/model/UpdateReservationRequest.kt
package com.example.smartsolarmicrogridtradingsystem.feature.reservationqr.model

import org.json.JSONObject

/**
 * Request payload for modifying a scheduled energy slot reservation via PUT /api/reservations/{id}.
 * Note: Per backend contract, station changes are not allowed on existing reservations.
 *
 * @property reservationDateTime Updated reservation timestamp in ISO-8601 format.
 * @property slotId Optional updated slot identifier.
 * @property reservationType Optional updated operational type (e.g. "DropOff", "Charging").
 */
data class UpdateReservationRequest(
    val reservationDateTime: String,
    val slotId: String? = null,
    val reservationType: String? = null
) {
    /**
     * Serializes this request into a [JSONObject] matching backend C# property conventions.
     */
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("reservationDateTime", reservationDateTime)
            if (!slotId.isNullOrBlank()) {
                put("slotId", slotId)
            }
            if (!reservationType.isNullOrBlank()) {
                put("reservationType", reservationType)
            }
        }
    }
}
